### PR TITLE
fix(moduleInstance): more careful evaluation order and error handling

### LINKED ISSUE #43: Modules not evaluated in source code order
**Steps to reproduce**
In `make-importer`, replace `moddir/index.js` with:
```js
// index.js
const global = Function('return this;')();
if (global.test262 !== '123456789') {
  throw `modules not evaluated in source code order ${global.test262}`;
}

import {} from './mod1';
import './mod2';
import * as ns1 from './mod3';
import dflt1 from './mod4';
export {} from './mod5';
import dflt2, {} from './mod6';
export * from './mod7';
import dflt3, * as ns2 from './mod8';
export * as ns3 from './mod9';
```
And create the following files
```js
// mod1.js
Function('return this;')().test262 = '1';
```
```js
// mod2.js
Function('return this;')().test262 += '2';
```
```js
// mod4.js
Function('return this;')().test262 += '3';
export default null;
```
```js
// mod4.js
Function('return this;')().test262 += '4';
export default null;
```
```js
// mod5.js
Function('return this;')().test262 += '5';
```
```js
// mod6.js
Function('return this;')().test262 += '6';
export default null;
```
```js
// mod7.js
Function('return this;')().test262 += '7';
```
```js
// mod8.js
Function('return this;')().test262 += '8';
export default null;
```
```js
// mod9.js
Function('return this;')().test262 += '9';
```

Actual result:
```
$ node -r esm test/test-moddir.js 
TAP version 13
# moddir index.js
not ok 1 unexpected exception
  ---
    operator: notEqual
    expected: 'modules not evaluated in source code order 132456789'
    actual:   'modules not evaluated in source code order 132456789'
 ```
Expected result, same as node's native import:
```sh
$ node -r esm test/moddir/index.js
// no error
```

### LINKED ISSUE #55: Modules not evaluated in source code order
**Steps to reproduce**
In `make-importer`, replace `moddir/index.js` with:
```js
// index.js
const global = Function('return this;')();
if (global.test262 !== '123456789') {
  throw `modules not evaluated in source code order ${global.test262}`;
}

import {} from './mod1';
import './mod2';
import * as ns1 from './mod3';
import dflt1 from './mod4';
export {} from './mod5';
import dflt2, {} from './mod6';
export * from './mod7';
import dflt3, * as ns2 from './mod8';
export * as ns3 from './mod9';
```
And create the following files
```js
// mod1.js
Function('return this;')().test262 = '1';
```
```js
// mod2.js
Function('return this;')().test262 += '2';
```
```js
// mod4.js
Function('return this;')().test262 += '3';
export default null;
```
```js
// mod4.js
Function('return this;')().test262 += '4';
export default null;
```
```js
// mod5.js
Function('return this;')().test262 += '5';
```
```js
// mod6.js
Function('return this;')().test262 += '6';
export default null;
```
```js
// mod7.js
Function('return this;')().test262 += '7';
```
```js
// mod8.js
Function('return this;')().test262 += '8';
export default null;
```
```js
// mod9.js
Function('return this;')().test262 += '9';
```

Actual result:
```
$ node -r esm test/test-moddir.js 
TAP version 13
# moddir index.js
not ok 1 unexpected exception
  ---
    operator: notEqual
    expected: 'modules not evaluated in source code order 132456789'
    actual:   'modules not evaluated in source code order 132456789'
 ```
Expected result, same as node's native import:
```sh
$ node -r esm test/moddir/index.js
// no error
```

### PR DESCRIPTION
**Pull request by michaelfig**
Closes #43

### CODE REVIEW

File: test/test-test262.js [hunk 1]
```diff
@@ -138,7 +138,7 @@ function hasExcludedInfo(testInfo) {
       // 'References observe the initialization of lexical bindings', // https://github.com/Agoric/make-importer/issues/40
       // 'References observe the mutation of initialized bindings', // https://github.com/Agoric/make-importer/issues/40
 
-      'Requested modules are evaluated prior to the requesting module in source code order', // https://github.com/Agoric/make-importer/issues/43
+      // 'Requested modules are evaluated prior to the requesting module in source code order', // https://github.com/Agoric/make-importer/issues/43
```
Comment (dckc) [inline test/test-test262.js:119]: why is this commented out?

Comment (michaelfig) [reply test/test-test262.js:119]: It's a test exclusion list.  Commenting it out allows the test to run.

**Comment by jfparadis** [comment]
Hi Michale, could you reverse. There is a problem.

File: test/test-test262.js [hunk 2]
```diff
@@ -192,6 +170,10 @@ function injectTest262Harness(globalObject, t) {
     throw 'Test262: This statement should not be evaluated.';
   }
 
+  // This cleans up (some of) the global state between tests.
+  // TODO: Have proper isolation.
+  delete globalObject.test262;
+
```
Comment (jfparadis) [inline test/test-test262.js:176]: We should not fix the test by doing something like that. This masquerade a valid problem with the system. This is not part of the official test262 harness.

File: test/test-test262.js [hunk 3]
```diff
@@ -97,19 +97,6 @@ function hasExcludedInfo(testInfo) {
       // some tests fail because they rely on a side effect which is failing
       // resolving an issue might uncover another issue
 
-      // 'Statement cannot contain an `import` declaration', // https://github.com/Agoric/transform-module/issues/5
```
Comment (jfparadis) [inline test/test-test262.js:100]: Let's keep these commented items, at least until we are out of the experimental phase. They are useful when debugging (you can uncomment a few at a time).

Comment (michaelfig) [reply test/test-test262.js:100]: Reverted.

File: test/test-test262.js (hunk 2 above)
Comment (michaelfig) [reply test/test-test262.js:176]: Reverted.

**Pull request merged**